ui/src/app/store/manage-collectiom/reducers.ts

```diff
@@ -56,7 +76,139 @@ const reducer: ActionReducer<CollectionEntityState> = createReducer(
+
+	  const newCollectionItems = items.map(item => ({
+		  url: CollectionHelper.getCollectionItemUrl(apiKey, item.image_id),
+		  id: item.image_id,
+		  status: CircleLoadingProgressEnum.Uploaded,
+		  subject: item.subject
+	  }));
+
+	  return {
+		  ...state,
+		  isCollectionPending: false,
+		  collection:[...newCollectionItems, ...collectionCopy]
+	  };
+  }),
+  on(getSubjectExamplesFail, (state) => ({...state, isCollectionPending: false})),
+  on(addFileToCollection, (state, {url, file, subject}) => ({
+	  ...state,
+	  collection: [
+		  ...state.collection,
+		  {url, file, subject, status: CircleLoadingProgressEnum.OnHold}
+		]
+	})
+  ),
+  on(uploadImage, (state, {item}) => {
+	const collectionCopy = [...state.collection];
+	const targetItemIndex = collectionCopy.findIndex((collectionCopyItem) => item.url === collectionCopyItem.url);
+
+	if (~targetItemIndex) {
+		collectionCopy[targetItemIndex] = {
+			...collectionCopy[targetItemIndex],
+			status: CircleLoadingProgressEnum.InProgress
+		}
+	}
+
+	return ({
+	  ...state,
+	  collection: collectionCopy
+	});
+  }),
+  on(uploadImageSuccess, (state, {item}) => {
+	const collectionCopy = [...state.collection];
+	const targetItemIndex = collectionCopy.findIndex((collectionCopyItem) => item.url === collectionCopyItem.url);
+
+	if (~targetItemIndex) {
+		collectionCopy[targetItemIndex] = {
+			...collectionCopy[targetItemIndex],
+			status: CircleLoadingProgressEnum.Uploaded
+		}
+	}
+
+	return ({
+		...state,
+		collection: collectionCopy
+	});
+  }),
+  on(uploadImageFail, (state, {item, error}) => {
+	const collectionCopy = [...state.collection];
+	const targetItemIndex = collectionCopy.findIndex((collectionCopyItem) => item.url === collectionCopyItem.url);
+
+	if (~targetItemIndex) {
+		collectionCopy[targetItemIndex] = {
+			...collectionCopy[targetItemIndex],
+			status: CircleLoadingProgressEnum.Failed,
+			error
+		}
+	}
+
+	return ({
+		...state,
+		collection: collectionCopy
+	});
+  }),
+  on(deleteSubjectExample, (state, {item}) => {
+	const collectionCopy = [...state.collection];
+	const targetItemIndex = collectionCopy.findIndex((collectionCopyItem) => item.url === collectionCopyItem.url);
+
+	if (~targetItemIndex) {
+		collectionCopy[targetItemIndex] = {
+			...collectionCopy[targetItemIndex],
+			status: CircleLoadingProgressEnum.InProgress
+		}
+	}
+
+	return ({
+		...state,
+		collection: collectionCopy
+	});
+  }),
+  on(deleteSubjectExampleSuccess, (state, {item}) => {
+	const collectionCopy = [...state.collection];
+	const targetItemIndex = collectionCopy.findIndex((collectionCopyItem) => item.url === collectionCopyItem.url);
+
+	if (~targetItemIndex) {
+		collectionCopy[targetItemIndex] = {
+			...collectionCopy[targetItemIndex],
+			status: CircleLoadingProgressEnum.Uploaded
+		}
+	}
+
+	return ({
+		...state,
+		collection: collectionCopy
+	});
+  }),
+  on(deleteSubjectExampleFail, (state, {item}) => {
+	const collectionCopy = [...state.collection];
```

Comment:
overall code of these <action> <actionSuccess> <actionFail> reducers is very similar and differs only by status enum - can be extracted to helper functon

Reply:
moved such logic to updateCollectionItemStatus function